Attempt to avoid repeated finds during housekeeping

ballard26 opened the pull request:
## Cover letter

This PR changes `log_manager::housekeeping` to attempt to iterate over `_logs` in one pass while looking for non-compacted logs. The current implementation will iterate over `_logs` twice per loop to find non-compacted logs. 

## Release notes
* none

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -127,31 +130,49 @@ ss::future<> log_manager::housekeeping() {
      */
+    std::unique_ptr<model::ntp> current_key{};
     using bflags = log_housekeeping_meta::bitflags;
-    return ss::do_until(
```
since nearly this entire function is being refactored, it might make sense to switch over to coroutines.

ballard26 replied:
Went ahead and refactored it into a coroutine. Thanks!

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -127,31 +130,49 @@ ss::future<> log_manager::housekeeping() {
-                       return it->second.handle.compact(compaction_config(
+    return ss::repeat([this,
+                       collection_threshold,
+                       current_key = std::move(current_key)]() mutable {
```
it is generally best practice to use `ss::do_with` to hold state that is going to live across loop iterations. for example

`ss::do_with(Type t, [](Type& instance) { return ss::repeat([&instance, ....`

although if you switch to coroutines then this state can live on the stack.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -127,31 +130,49 @@ ss::future<> log_manager::housekeeping() {
+                     typename logs_type::iterator it, start_it;
+
+                     if (!current_key) {
+                         current_key.reset(new model::ntp());
```
you might consider switching to `std::optional<model::ntp>` to express your intent more explicitly where `current_key` is initially unset. i'm not seeing a reason it has to be a unique_ptr on the heap.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -127,31 +130,49 @@ ss::future<> log_manager::housekeeping() {
+                         current_key.reset(new model::ntp());
+                         start_it = _logs.begin();
+                     } else {
+                         start_it = _logs.find(*current_key);
+                     }
+
+                     if (start_it != _logs.end()) {
+                         it = find_next_non_compacted_log(_logs, start_it);
+                     } else {
+                         // The ntp was deleted.
+                         it = find_next_non_compacted_log(_logs, _logs.begin());
```
starting at the beginning when the current_key isn't found seems completely reasonable. i am wondering though if it would make sense to use something like `_logs.lower_bound` or `_logs.upper_bound` such that if current_key were to be removed, then the search would start at whatever would have come before current_key. so it might result in some duplicates, but not necessarily from the very beginning. again just thinking about this; there may be some cases i'm not thinking about.

ballard26 commented:
Thanks for the feedback! I went ahead and refactored part of the function into a coroutine. Also I've changed some of the logic used to detect if the iterator we're using to scan `_logs` is still valid after seastar reschedules the scan. This was done since there was a slight issue with keeping a key and using that to re-establish a valid iterator. If `_logs` was rehashed between the loop iterations then although the key would still be in `_logs` it would be in a new random position that would lead to us missing log entries to compact. The new logic instead adds a counter `_logs_version` that is incremented every time `_logs` is modified. That way we can detect rehashes of `_logs` as well. 

If everything looks fine with you folks I'll go ahead an refactored the commit log into 2 parts. One for the refactor to coroutines and the other for the logic change.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -94,14 +95,55 @@ ss::future<> log_manager::stop() {
 }
 
-static inline logs_type::iterator find_next_non_compacted_log(logs_type& logs) {
+inline logs_type& log_manager::modify_logs() {
+    _logs_version++;
+    return _logs;
+}
```
clever 😎

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -94,14 +95,55 @@ ss::future<> log_manager::stop() {
+          _config.compaction_priority,
+          _abort_source));
+
+        if (current_version != _logs_version) {
```
props on detecting changes here. we do something similar in the memory claimer and generally it works really well.

the downside to this is that it's prone to developer mistakes (e.g. not going through the modify() wrapper). but unlike the memory reclaimer, this loop doesn't seem to be in the hot path, it merely needs to be efficient for large collections. so i'm wondering if there are other approaches, here are a couple examples:

1. what if we kept the `_logs` collection the way it is now as the source of true, but constructed a secondary index as a separate collection of logs that marked as needing to be compacted. in the simple case we'd initialize both containers to be the same, and pop an element out of secondary index to compact it. when the container is empty we'd scan the source of truth to repopulate.

2. another option would be to store the logs in a boost instructive list ordered by least-recently-compacted. the O(1) index from ntp->log would still be used, but each time we compacted we'd stick the compated log at the head of the intrusive list (very fast).

I'm not opposed to this particular trick with the version, but I'm just curious if there are other points in the design space worth thinking about.

ballard26 replied:
We can avoid developer mistakes by doing something like;
```cpp
class logs_protected {
     logs_type _logs;
public:
    logs_type& modify_logs();
    logs_type const& read_logs();
};
```

It the space complexity isn't an issue keeping a separate queue of logs from the `_logs` hash map sounds like a good idea. We could add the log to the queue at the same time we add it to `_logs`. Then we could remove a set number per housekeeping and then re-add them to the back of the queue after compaction.

dotnwat replied:
>     logs_type const& read_logs();

i can still obtain an iterator from this interface, hold on to it, and get myself it trouble. but it's better than nothing!

> However, removing logs would still invalidate iterators. 

i haven't thought about it in detail. but i think the key is to not hold on to any iterators. either by looking up by ntp in both containers, or representing the relationships explicitly.

One option is something like this:

```c++
struct log_wrapper {
   intrusive_list_link;
   log l;
};

map<ntp, std::unique_ptr<log_wrapper>> logs;
intrusive_list<log_wrapper>> least_recently_compacted;
```

* to find a a log by ntp, use logs[ntp]
* to find the next candidate to compact, look at the first element on the lrc list
* when a compaction target is chosen, move it to the end of the list
* when a log is deleted, delete it from both lists

the key thing i'm assuming here is that the least-recently-compacted organization provides the search ordering that you want. if not, i may be leading you down a bad path here.

do you or @jcsp have a target number of elements that will be in the container? the thing is that absl::flat_hash_map may cause problems for large numbers of partitions. but we can swap in absl::btree_map in this case. intrusive list doesn't have any sizing problems related to memory allocations.

ballard26 replied:
Went ahead and implemented the intrusive list version of this. As you've said it definitely reduces potential developer errors in the future. I think we have some goals for total size related to how many partitions we can run on a single node. But that size is nowhere's near `absl::flat_hash_map` being an issue.

ballard26 replied:
Unfortunately intrusive lists don't work well when storing their nodes in a container that dynamically reallocates it's self. I've had to switch to a `std::list` unfortunately. This means there is a lot more memory fragmentation though unfortunately along with additional logic to keep `_logs` and `_logs_list` in sync with one another. Let me know what you think.

dotnwat commented:
just wanted to echo what john said about composing a PR that does a conversion to coroutines:

because the conversion to coroutines often means that nearly the entire block of codes is deleted and added from a diff point-of-view, the goal is to make reviewing that the changes preserve the original semantics. a series of commits as follows usually works as a template:

* commits 1-N: conversion of old code to co-routines
  * this might be one commit or it is often like 1 commit per level of nested continuation
  * it's fine to disable clang-format for these commits
  * maximize for understandability
* commit N+1: apply clang-format
* commits N+2... the original changes now in coroutine-space

dotnwat commented on src/v/storage/log_housekeeping_meta.h:
```diff
@@ -13,8 +13,14 @@
+
 namespace storage {
 struct log_housekeeping_meta {
+    using intrusive_link_type = boost::intrusive::list_member_hook<
+      boost::intrusive::link_mode<boost::intrusive::auto_unlink>>;
```
we have a header `utils/intrusive_list_helpers.h` that provide typedefs for common cases that we use in the tree.

dotnwat commented on src/v/storage/log_manager.h:
```diff
@@ -227,7 +228,15 @@ class log_manager {
-    using logs_type = absl::flat_hash_map<model::ntp, log_housekeeping_meta>;
+    using logs_type
+      = absl::flat_hash_map<model::ntp, std::unique_ptr<log_housekeeping_meta>>;
+    using compaction_list_type = boost::intrusive::list<
+      log_housekeeping_meta,
+      boost::intrusive::member_hook<
+        log_housekeeping_meta,
+        log_housekeeping_meta::intrusive_link_type,
+        &log_housekeeping_meta::link>,
+      boost::intrusive::constant_time_size<false>>;
```
i think using the typedefs in `utils/intrusive_list_helpers.h` will simplify this typedef a bit

dotnwat replied:
here is an example from `batch_cache.h`

```c++
intrusive_list<range, &range::_hook> _lru;
```

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -88,18 +88,44 @@ ss::future<> log_manager::stop() {
+
+        // Make sure the current log hasn't been removed
+        // from the _logs_list.
+        if (&current_log == &_logs_list.front()) {
```
with very few exceptions we should really try hard to avoid low-level things stuff like comparing memory raw memory addresses to do high-level operations like test for container membership.

can we avoid this by moving the choice to the back of list before we call co_await? then it is is in the correct position, and if any other operation in the system removed the log it would be responsible for removing the entry from both indicies (which i assume is already the case)

ballard26 replied:
It shouldn't be an issue moving the list repositioning to before the `co_await`. Then the memory address won't be needed. Good call.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -88,18 +88,44 @@ ss::future<> log_manager::stop() {
+        co_return;
+    }
+
+    while ((_logs_list.front().flags & bflags::compacted) == bflags::none) {
```
is there a possibility of a scenario occurring where the first item doesn't need to be compacted, but the second element does and this loop won't run and find the second element? this seems very error prone in order to maintain this property. perhaps we should still scan the whole list? the alternative is to maintain data structures that serve as indicies over exactly the type of query we want to execute (e.g. GC/retention time is due, etc...)

ballard26 replied:
The bflags::compacted might be a misnomer in this case. It doesn't really indicate if the log has been compacted or not. I believe that the logic to determine if the log should be compacted is in the log's compact function. The log_manager is just responsible for ensuring that function is called periodically on every log.

To that end the compacted flag is cleared on all logs before the housekeeping_scan starts then it is set one at a time as compacted is called on the log. So then the housekeeping_scan reaches a log where the flag has already been set then we know we've called compact on every log.

So in essence we are scanning the entire list each time. But rather than holding an iterator we are marking a flag on each element we read.

dotnwat replied:
Oh of course. I see at the end of housekeeping all the flags get cleared again

```
      .finally([this] {
          for (auto& h : _logs) {
              h.second.flags &= ~bflags::compacted;
          for (auto& h : _logs_list) {
              h.flags &= ~bflags::compacted;
          }
```

thanks for reminding me!

dotnwat replied:
ill resolve this it isn't an issue at all.

dotnwat reviewed (commented):
looking good. i left a few comments/questions inline.

1. it looks like your second commit may replace the first. since you've changed the strategy of your solution, it would make sense to squash the second commit into the first.
2. would you mind adding to the PR coverletter (or ideally the commit message) an explanation of the problem we are solving and how this PR solves it? I think one challenge with this sort of change is that the design space is large so keeping the goal in mind would be useful.

dotnwat reviewed (dismissed):
lgtm. nice work. just waiting on CI to finish before merging

ballard26 commented on src/v/storage/log_manager.cc:
```diff
@@ -248,7 +237,9 @@ ss::future<log> log_manager::do_manage(ntp_config cfg) {
               auto l = storage::make_disk_backed_log(
                 std::move(cfg), *this, std::move(segments), _kvstore);
-              auto [_, success] = _logs.emplace(l.config().ntp(), l);
+              auto [_, success] = _logs.emplace(
+                l.config().ntp(), std::make_unique<log_housekeeping_meta>(l));
+              _logs_list.push_back(*_logs.at(l.config().ntp()));
```
Good catch, making the change now.

dotnwat commented:
Looks like a memory leak

```


=================================================================
--
  | ==1458==ERROR: LeakSanitizer: detected memory leaks
  |  
  | Direct leak of 88 byte(s) in 1 object(s) allocated from:
  | #0 0x5632e176b69d  (/var/lib/buildkite-agent/builds/buildkite-amd64-builders-i-0faacf2b3e9914be3-1/redpanda/redpanda/vbuild/debug/clang/bin/test_cluster_rpunit+0x30a3c69d) (BuildId: 929666b60085e11661fecb9488f1ff63b627a926)
  | #1 0x5632f1453378  (/var/lib/buildkite-agent/builds/buildkite-amd64-builders-i-0faacf2b3e9914be3-

```

Might be unrelated, but i haven't seen anything like this recently. Check existing issues if others have reported.

I'm surprised that the stack trace wasn't decoded automatically. Perhaps it is related to the recent issue reported with clang-14 which Travis D reported.
